File: pe-math/src/main/java/org/saar/maths/noise/LayeredNoise3f.java
package org.saar.maths.noise;

public class LayeredNoise3f implements Noise3f {

    private final Noise3f noise3f;
    private final int layers;

    public LayeredNoise3f(Noise3f noise3f, int layers) {
        this.noise3f = noise3f;
        this.layers = layers;
    }

    @Override
    public float noise(float x, float y, float z) {
        float noise = 0f;
        for (int i = 0; i < this.layers; i++) {
            final float pow2 = (float) Math.pow(2, i);
            noise += this.noise3f.noise(x / pow2, y / pow2, z / pow2) * pow2;
        }
        return (float) (noise / (Math.pow(2, this.layers) - 1));
    }
}

File: pe-math/src/main/kotlin/org/saar/maths/noise/LayeredNoise3f.kt
package org.saar.maths.noise

import kotlin.math.pow

class LayeredNoise3f(private val noise3f: Noise3f, private val layers: Int) : Noise3f {

    override fun noise(x: Float, y: Float, z: Float): Float {
        val noise = (0 until this.layers).sumOf {
            val pow2 = 2.0.pow(it.toDouble()).toFloat()
            (this.noise3f.noise(x / pow2, y / pow2, z / pow2) * pow2).toDouble()
        }
        return (noise / (2.0.pow(this.layers.toDouble()) - 1)).toFloat()
    }
}
